RFC: no default link attributes when marshalling

This is a request for comments. This library sets various default attributes when serializing a LinkMessage. I am no netlink expert but I noticed that this leads to errors when, for example, bringing an interface up or down, specifically `netlink receive: invalid argument`. I observed that `ip` doesn't send any attribute, and that's the only difference between a successful and an unsuccessful run. The `strace` output from the sample code below compares the output produced by `ip link set eth0 up` vs. the Go program at the end. The only difference is the presence of various attributes.
Furthermore, these attributes are serialized from the zero-value of the `LinkAttributes` structure, which leads to erroneous values.
I am proposing here to change the default behaviour to *not* serialize attributes if none is explicitly specified. With such change, the program below succeeds.

## Strace output
Using `ip`:
```
$ sudo strace -esendmsg ip link set wlan0 up
sendmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base={{len=32, type=RTM_NEWLINK, flags=NLM_F_REQUEST|NLM_F_ACK, seq=1554819541, pid=0}, {ifi_family=AF_UNSPEC, ifi_type=ARPHRD_NETROM, ifi_index=if_nametoindex("wlan0"), ifi_flags=IFF_UP, ifi_change=0x1}}, iov_len=32}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 32
+++ exited with 0 +++
```
while using this library I get:
```
$ go build && sudo strace -esendmsg ./example wlan0
sendmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base={{len=72, type=RTM_SETLINK, flags=NLM_F_REQUEST|NLM_F_ACK, seq=2960429762, pid=16588}, {ifi_family=AF_UNSPEC, ifi_type=ARPHRD_NETROM, ifi_index=if_nametoindex("wlp2s0"), ifi_flags=0, ifi_change=0x1}, [{{nla_len=6, nla_type=IFLA_UNSPEC}, "\x00\x00"}, {{nla_len=5, nla_type=IFLA_IFNAME}, ""}, {{nla_len=8, nla_type=IFLA_MTU}, 0}, {{nla_len=8, nla_type=IFLA_LINK}, 0}, {{nla_len=5, nla_type=IFLA_QDISC}, ""}]}, iov_len=72}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 72
2019/04/09 15:44:08 netlink receive: invalid argument
panic: netlink receive: invalid argument
```

## Code of the failing program

```go
package main

import (
	"log"
	"net"

	"github.com/jsimonetti/rtnetlink"
	"golang.org/x/sys/unix"
)

func main() {
	conn, err := rtnetlink.Dial(nil)
	if err != nil {
		log.Panic(err)
	}
	iface, err := net.InterfaceByName("wlan0")
	if err != nil {
		log.Panic(err)
	}
	err = conn.Link.Set(&rtnetlink.LinkMessage{
		Family: unix.AF_UNSPEC,
		Type:   unix.ARPHRD_NETROM,
		Index:  uint32(iface.Index),
		Flags:  unix.IFF_UP,
		Change: unix.IFF_UP,
	})
	if err != nil {
		log.Panic(err)
	}
}
```

--- a/conn_test.go
+++ b/conn_test.go
@@ -3,7 +3,6 @@ package rtnetlink
 import (
 	"encoding"
 	"fmt"
-	"net"
 	"os"
 	"reflect"
 	"testing"
@@ -30,15 +29,6 @@ func TestConnExecute(t *testing.T) {
 			Index:  0x4030201,
 			Flags:  0x101,
 			Change: 0x4030201,
-			Attributes: LinkAttributes{
-				Address:   net.HardwareAddr(nil),
-				Broadcast: net.HardwareAddr(nil),
-				Name:      "",
-				MTU:       0x0,
-				Type:      0x0,
-				QueueDisc: "",
-				Stats:     nil,
-			},
 		},
 	}
 
@@ -147,15 +137,6 @@ func TestConnReceive(t *testing.T) {
 			Index:  0x0,
 			Flags:  0x0,
 			Change: 0x0,
-			Attributes: LinkAttributes{
-				Address:   net.HardwareAddr(nil),
-				Broadcast: net.HardwareAddr(nil),
-				Name:      "",
-				MTU:       0x0,
-				Type:      0x0,
-				QueueDisc: "",
-				Stats:     nil,
-			},
 		},
 		{
 			Family: 0x102,
@@ -163,15 +144,6 @@ func TestConnReceive(t *testing.T) {
 			Index:  0x4030201,
 			Flags:  0x102,
 			Change: 0x4030201,
-			Attributes: LinkAttributes{
-				Address:   net.HardwareAddr(nil),
-				Broadcast: net.HardwareAddr(nil),
-				Name:      "",
-				MTU:       0x0,
-				Type:      0x0,
-				QueueDisc: "",
-				Stats:     nil,
-			},
 		},
 	}
 

--- a/link.go
+++ b/link.go
@@ -39,7 +39,7 @@ type LinkMessage struct {
 	Change uint32
 
 	// Attributes List
-	Attributes LinkAttributes
+	Attributes *LinkAttributes
 }
 
 const linkMessageLength = 16
@@ -55,12 +55,16 @@ func (m *LinkMessage) MarshalBinary() ([]byte, error) {
 	nlenc.PutUint32(b[8:12], m.Flags)
 	nlenc.PutUint32(b[12:16], m.Change)
 
-	a, err := m.Attributes.MarshalBinary()
-	if err != nil {
-		return nil, err
+	if m.Attributes != nil {
+		a, err := m.Attributes.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+
+		return append(b, a...), nil
 	}
 
-	return append(b, a...), nil
+	return b,nil
 }
 
 // UnmarshalBinary unmarshals the contents of a byte slice into a LinkMessage.
@@ -77,7 +81,7 @@ func (m *LinkMessage) UnmarshalBinary(b []byte) error {
 	m.Change = nlenc.Uint32(b[12:16])
 
 	if l > linkMessageLength {
-		m.Attributes = LinkAttributes{}
+		m.Attributes = &LinkAttributes{}
 		err := m.Attributes.UnmarshalBinary(b[16:])
 		if err != nil {
 			return err

--- a/link_test.go
+++ b/link_test.go
@@ -22,11 +22,6 @@ func TestLinkMessageMarshalBinary(t *testing.T) {
 			b: []byte{
 				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x06, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x05, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x08, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x08, 0x00, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x05, 0x00, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00,
 			},
 		},
 		{
@@ -41,11 +36,6 @@ func TestLinkMessageMarshalBinary(t *testing.T) {
 			b: []byte{
 				0x00, 0x00, 0x01, 0x00, 0x02, 0x00, 0x00, 0x00,
 				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x06, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x05, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x08, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x08, 0x00, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x05, 0x00, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00,
 			},
 		},
 		{
@@ -60,17 +50,12 @@ func TestLinkMessageMarshalBinary(t *testing.T) {
 			b: []byte{
 				0x00, 0x00, 0x01, 0x00, 0x02, 0x00, 0x00, 0x00,
 				0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x06, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x05, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x08, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x08, 0x00, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00,
-				0x05, 0x00, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00,
 			},
 		},
 		{
 			name: "attributes",
 			m: &LinkMessage{
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:   []byte{0x40, 0x41, 0x42, 0x43, 0x44, 0x45},
 					Broadcast: []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
 					Name:      "lo",
@@ -92,7 +77,7 @@ func TestLinkMessageMarshalBinary(t *testing.T) {
 		{
 			name: "attributes ipip",
 			m: &LinkMessage{
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:   []byte{10, 0, 0, 1},
 					Broadcast: []byte{255, 255, 255, 255},
 					Name:      "ipip",
@@ -114,7 +99,7 @@ func TestLinkMessageMarshalBinary(t *testing.T) {
 		{
 			name: "info",
 			m: &LinkMessage{
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:   []byte{0, 0, 0, 0, 0, 0},
 					Broadcast: []byte{0, 0, 0, 0, 0, 0},
 					Name:      "lo",
@@ -149,7 +134,7 @@ func TestLinkMessageMarshalBinary(t *testing.T) {
 		{
 			name: "operational state",
 			m: &LinkMessage{
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:          []byte{10, 0, 0, 1},
 					Broadcast:        []byte{255, 255, 255, 255},
 					Name:             "ipip",
@@ -235,7 +220,7 @@ func TestLinkMessageUnmarshalBinary(t *testing.T) {
 				0x05, 0x00, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00,
 			},
 			m: &LinkMessage{
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:   []byte{0, 0, 0, 0, 0, 0},
 					Broadcast: []byte{0, 0, 0, 0, 0, 0},
 				},
@@ -261,7 +246,7 @@ func TestLinkMessageUnmarshalBinary(t *testing.T) {
 				Index:  2,
 				Flags:  0,
 				Change: 0,
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:   []byte{0, 0, 0, 0, 0, 0},
 					Broadcast: []byte{0, 0, 0, 0, 0, 0},
 				},
@@ -282,7 +267,7 @@ func TestLinkMessageUnmarshalBinary(t *testing.T) {
 				0x05, 0x00, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00,
 			},
 			m: &LinkMessage{
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:   []byte{0, 0, 0, 0, 0, 0},
 					Broadcast: []byte{0, 0, 0, 0, 0, 0},
 					Name:      "lo",
@@ -304,7 +289,7 @@ func TestLinkMessageUnmarshalBinary(t *testing.T) {
 				0x00, 0x00, 0x00, 0x00,
 			},
 			m: &LinkMessage{
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:   []byte{10, 0, 0, 1},
 					Broadcast: []byte{255, 255, 255, 255},
 					Name:      "ipip",
@@ -333,7 +318,7 @@ func TestLinkMessageUnmarshalBinary(t *testing.T) {
 				0x05, 0x06, 0x07, 0x08, 0x09, 0x00, 0x00, 0x00,
 			},
 			m: &LinkMessage{
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:   []byte{0, 0, 0, 0, 0, 0},
 					Broadcast: []byte{0, 0, 0, 0, 0, 0},
 					Name:      "lo",
@@ -362,7 +347,7 @@ func TestLinkMessageUnmarshalBinary(t *testing.T) {
 				0x06, 0x00, 0x00, 0x00,
 			},
 			m: &LinkMessage{
-				Attributes: LinkAttributes{
+				Attributes: &LinkAttributes{
 					Address:          []byte{10, 0, 0, 1},
 					Broadcast:        []byte{255, 255, 255, 255},
 					Name:             "ipip",

